{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Additional information about the PAO1 and TCGA pathway-pathway networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import itertools\n",
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import fisher_exact\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "data_dir = os.path.join(\"..\", \"data\")\n",
    "pao1_data = os.path.join(data_dir, \"pao1_data\")\n",
    "tcga_data = os.path.join(data_dir, \"tcga_data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pao1_kegg_edges_file = os.path.join(\n",
    "    pao1_data, \"eADAGE_analysis\", \"permutation_test_n=10000\", \"filtered_network.tsv\")\n",
    "tcga_pid_edges_file = os.path.join(\n",
    "    tcga_data, \"NMF_analysis\", \"permutation_test_n=10000\", \"filtered_network.tsv\")\n",
    "\n",
    "kegg_pathways = os.path.join(pao1_data, \"pseudomonas_KEGG_terms.txt\")\n",
    "pid_pathways = os.path.join(tcga_data, \"PID_pathway_definitions.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def load_pathway_definitions_file(pathways_file, shorten_pathway_name):\n",
    "    pathway_definitions_df = pd.read_table(\n",
    "        pathways_file, header=None, names=[\"pw\", \"size\", \"genes\"])\n",
    "    pathway_definitions_df[\"genes\"] = pathway_definitions_df[\"genes\"].map(\n",
    "        lambda x: x.split(\";\"))\n",
    "    pathway_definitions_df.set_index(\"pw\", inplace=True)\n",
    "\n",
    "    pathway_definitions_map = {}\n",
    "    for index, row in pathway_definitions_df.iterrows():\n",
    "        pathway = shorten_pathway_name(index)\n",
    "        pathway_definitions_map[pathway] = set(row[\"genes\"])\n",
    "    return pathway_definitions_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pw0</th>\n",
       "      <th>pw1</th>\n",
       "      <th>weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ribosome PAO1</td>\n",
       "      <td>Ribosome, bacteria</td>\n",
       "      <td>26.089640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Phosphonate and phosphinate metabolism PAO1</td>\n",
       "      <td>Phosphonate transport system</td>\n",
       "      <td>25.455394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Macrolide resistance, MacAB-TolC transporter</td>\n",
       "      <td>Zinc transport system</td>\n",
       "      <td>23.170151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Phosphonate and phosphinate metabolism PAO1</td>\n",
       "      <td>Type II general secretion pathway</td>\n",
       "      <td>21.168310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Glycine betaine/proline transport system</td>\n",
       "      <td>Glycine, serine and threonine metabolism PAO1</td>\n",
       "      <td>19.221697</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            pw0  \\\n",
       "0                                 Ribosome PAO1   \n",
       "1   Phosphonate and phosphinate metabolism PAO1   \n",
       "2  Macrolide resistance, MacAB-TolC transporter   \n",
       "3   Phosphonate and phosphinate metabolism PAO1   \n",
       "4      Glycine betaine/proline transport system   \n",
       "\n",
       "                                             pw1     weight  \n",
       "0                             Ribosome, bacteria  26.089640  \n",
       "1                   Phosphonate transport system  25.455394  \n",
       "2                          Zinc transport system  23.170151  \n",
       "3              Type II general secretion pathway  21.168310  \n",
       "4  Glycine, serine and threonine metabolism PAO1  19.221697  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pao1_kegg_network = pd.read_table(pao1_kegg_edges_file)\n",
    "pao1_kegg_network.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# shorten the PAO1 KEGG pathway names so that they match the ones\n",
    "# we use in the publication\n",
    "def shorten_pao1_kegg(pathway_name):\n",
    "    REMOVE_SUFFIX = \"- Pseudomonas aeruginosa PAO1\"\n",
    "    pathway_short = None\n",
    "    split_label = pathway_name.split(\" \", 1)\n",
    "    if len(split_label) > 1:\n",
    "        pathway_short = split_label[1]\n",
    "    else:\n",
    "        pathway_short = split_label[0]\n",
    "    if REMOVE_SUFFIX in pathway_short:\n",
    "        remove_from_index = pathway_short.index(REMOVE_SUFFIX)\n",
    "        return \"{0}PAO1\".format(pathway_short[:remove_from_index])\n",
    "    return pathway_short.strip()\n",
    "\n",
    "kegg_pathways_dict = load_pathway_definitions_file(\n",
    "    kegg_pathways, shorten_pao1_kegg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PAO1 KEGG network summary statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the PAO1 KEGG pathway-pathway co-occurrence network:\n",
      "Number of distinct pathways in the network: 89\n",
      "Number of edges in the network: 203\n",
      "Number of edges in the network where the 2 pathways share genes: 35 (average similarity by Jaccard Index: 0.03482)\n"
     ]
    }
   ],
   "source": [
    "pathways_in_kegg_network = set(\n",
    "    pao1_kegg_network[\"pw0\"].tolist() + pao1_kegg_network[\"pw1\"].tolist())\n",
    "\n",
    "n_overlapping = 0\n",
    "edge_jaccards = []\n",
    "for index, row in pao1_kegg_network.iterrows():\n",
    "    pw0_genes = kegg_pathways_dict[row[\"pw0\"]] \n",
    "    pw1_genes = kegg_pathways_dict[row[\"pw1\"]]\n",
    "    overlap = pw0_genes & pw1_genes\n",
    "    if overlap:\n",
    "        n_overlapping += 1\n",
    "    jaccard = float(len(overlap)) / (len(pw0_genes) + len(pw1_genes) - len(overlap))\n",
    "    edge_jaccards.append(jaccard)        \n",
    "        \n",
    "print(\"For the PAO1 KEGG pathway-pathway co-occurrence network:\")\n",
    "print(\"Number of distinct pathways in the network: {0}\".format(\n",
    "    len(pathways_in_kegg_network)))\n",
    "print(\"Number of edges in the network: {0}\".format(\n",
    "    pao1_kegg_network.shape[0]))\n",
    "print(\"Number of edges in the network where the 2 pathways \"\n",
    "      \"share genes: {0} (average similarity by Jaccard Index: {1:.4})\".format(\n",
    "          n_overlapping, np.average(edge_jaccards)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison to a network based only on overlapping genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we construct a network only from gene overlap (FDR-corrected hypergeometric test < 0.05). \n",
    "The PathCORE-T network had 203 edges between 89 pathways. \n",
    "The overlap-based network (see below) had 406 edges between 158 pathways. \n",
    "Only 35 of the edges in the PathCORE-T network were between pathways that shared genes, with a low average Jaccard Index of 0.035. \n",
    "A network constructed using PathCORE-T (with overlap-correction by default) captures pathway co-occurrences not driven by shared genes between pathways. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def create_overlap_based_network(pathway_genes_map, alpha=0.05):\n",
    "    all_genes = set().union(*pathway_genes_map.values())\n",
    "    n_genes = len(all_genes)\n",
    "    \n",
    "    pvalues_list = []\n",
    "    pathway_pairs = list(itertools.combinations(\n",
    "        pathway_genes_map.keys(), 2))\n",
    "    for pw0, pw1 in pathway_pairs:\n",
    "        pw0_genes = pathway_genes_map[pw0]\n",
    "        pw1_genes = pathway_genes_map[pw1]\n",
    "        n_common = len(pw0_genes & pw1_genes)\n",
    "        n_pw0 = len(pw0_genes - pw1_genes)\n",
    "        n_pw1 = len(pw1_genes - pw0_genes)\n",
    "        n_all_other_genes = n_genes - n_pw0 - n_pw1 - n_common\n",
    "        contingency_table = np.array(\n",
    "            [[n_common, n_pw0],\n",
    "             [n_pw1, n_all_other_genes]])\n",
    "        _, pvalue = fisher_exact(\n",
    "            contingency_table, alternative=\"greater\")\n",
    "        pvalues_list.append(pvalue)\n",
    "    \n",
    "    below_alpha, qvalues, _, _ = multipletests(\n",
    "        pvalues_list, alpha=alpha, method=\"fdr_bh\")\n",
    "    \n",
    "    significant_edges = [(pathway_pairs[ix], qvalues[ix]) for ix, is_below \n",
    "                         in enumerate(below_alpha) if is_below]\n",
    "    return significant_edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of edges in the PAO1 KEGG overlap-based network: 406\n",
      "Number of pathways in overlap-based network: 158\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pw0</th>\n",
       "      <th>pw1</th>\n",
       "      <th>adjusted_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ribosome PAO1</td>\n",
       "      <td>Ribosome, bacteria</td>\n",
       "      <td>4.826194e-90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fatty acid metabolism PAO1</td>\n",
       "      <td>Fatty acid biosynthesis PAO1</td>\n",
       "      <td>5.252269e-41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Citrate cycle (TCA cycle) PAO1</td>\n",
       "      <td>Citrate cycle (TCA cycle, Krebs cycle)</td>\n",
       "      <td>6.427679e-40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bacterial secretion system PAO1</td>\n",
       "      <td>Type II general secretion pathway</td>\n",
       "      <td>4.308885e-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Citrate cycle, second carbon oxidation, 2-oxog...</td>\n",
       "      <td>Citrate cycle (TCA cycle, Krebs cycle)</td>\n",
       "      <td>2.499061e-33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Valine, leucine and isoleucine degradation PAO1</td>\n",
       "      <td>Propanoate metabolism PAO1</td>\n",
       "      <td>6.428247e-33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Biotin metabolism PAO1</td>\n",
       "      <td>Pimeloyl-ACP biosynthesis, BioC-BioH pathway, ...</td>\n",
       "      <td>1.409373e-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Fatty acid biosynthesis PAO1</td>\n",
       "      <td>Fatty acid biosynthesis, elongation</td>\n",
       "      <td>1.853222e-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Fatty acid degradation PAO1</td>\n",
       "      <td>Fatty acid metabolism PAO1</td>\n",
       "      <td>5.967364e-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Citrate cycle (TCA cycle) PAO1</td>\n",
       "      <td>Citrate cycle, second carbon oxidation, 2-oxog...</td>\n",
       "      <td>1.258273e-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Fatty acid degradation PAO1</td>\n",
       "      <td>Valine, leucine and isoleucine degradation PAO1</td>\n",
       "      <td>4.631031e-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Valine, leucine and isoleucine degradation PAO1</td>\n",
       "      <td>Leucine degradation, leucine =&gt; acetoacetate +...</td>\n",
       "      <td>1.873092e-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Aminoacyl-tRNA biosynthesis PAO1</td>\n",
       "      <td>Aminoacyl-tRNA biosynthesis, prokaryotes</td>\n",
       "      <td>3.641385e-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Fatty acid biosynthesis, elongation</td>\n",
       "      <td>Pimeloyl-ACP biosynthesis, BioC-BioH pathway, ...</td>\n",
       "      <td>3.641385e-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Fatty acid metabolism PAO1</td>\n",
       "      <td>Fatty acid biosynthesis, elongation</td>\n",
       "      <td>1.342633e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Fatty acid degradation PAO1</td>\n",
       "      <td>beta-Oxidation</td>\n",
       "      <td>1.862063e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Lysine biosynthesis PAO1</td>\n",
       "      <td>Lysine biosynthesis, succinyl-DAP pathway, asp...</td>\n",
       "      <td>2.719932e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Valine, leucine and isoleucine biosynthesis PAO1</td>\n",
       "      <td>Isoleucine biosynthesis, threonine =&gt; 2-oxobut...</td>\n",
       "      <td>3.267995e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Biotin metabolism PAO1</td>\n",
       "      <td>Fatty acid biosynthesis, elongation</td>\n",
       "      <td>2.512307e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Phenazine biosynthesis PAO1</td>\n",
       "      <td>Pyocyanine biosynthesis, chorismate =&gt; pyocyanine</td>\n",
       "      <td>4.340715e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Tryptophan metabolism PAO1</td>\n",
       "      <td>Lysine degradation PAO1</td>\n",
       "      <td>1.646385e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Fatty acid biosynthesis PAO1</td>\n",
       "      <td>Pimeloyl-ACP biosynthesis, BioC-BioH pathway, ...</td>\n",
       "      <td>3.174955e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Quorum sensing PAO1</td>\n",
       "      <td>Phenazine biosynthesis PAO1</td>\n",
       "      <td>5.955826e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Fatty acid metabolism PAO1</td>\n",
       "      <td>beta-Oxidation</td>\n",
       "      <td>2.398959e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Valine, leucine and isoleucine biosynthesis PAO1</td>\n",
       "      <td>2-Oxocarboxylic acid metabolism PAO1</td>\n",
       "      <td>4.864250e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Propanoate metabolism PAO1</td>\n",
       "      <td>Fatty acid metabolism PAO1</td>\n",
       "      <td>1.038392e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Valine, leucine and isoleucine degradation PAO1</td>\n",
       "      <td>beta-Alanine metabolism PAO1</td>\n",
       "      <td>1.144403e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Protein export PAO1</td>\n",
       "      <td>Sec (secretion) system</td>\n",
       "      <td>1.676715e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Folate biosynthesis PAO1</td>\n",
       "      <td>Tetrahydrofolate biosynthesis, GTP =&gt; THF</td>\n",
       "      <td>1.822326e-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Biotin metabolism PAO1</td>\n",
       "      <td>Fatty acid biosynthesis PAO1</td>\n",
       "      <td>2.184256e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Arginine biosynthesis PAO1</td>\n",
       "      <td>Alanine, aspartate and glutamate metabolism PAO1</td>\n",
       "      <td>4.212918e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Imipenem resistance, repression of porin OprD</td>\n",
       "      <td>CusS-CusR (copper tolerance) two-component reg...</td>\n",
       "      <td>8.324803e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Fatty acid degradation PAO1</td>\n",
       "      <td>Tryptophan metabolism PAO1</td>\n",
       "      <td>1.059795e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Valine, leucine and isoleucine biosynthesis PAO1</td>\n",
       "      <td>Valine/isoleucine biosynthesis, pyruvate =&gt; va...</td>\n",
       "      <td>1.397409e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Fatty acid degradation PAO1</td>\n",
       "      <td>Benzoate degradation PAO1</td>\n",
       "      <td>1.608228e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>DNA replication PAO1</td>\n",
       "      <td>Mismatch repair PAO1</td>\n",
       "      <td>2.628312e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>beta-Alanine metabolism PAO1</td>\n",
       "      <td>beta-Oxidation</td>\n",
       "      <td>2.898540e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Purine metabolism PAO1</td>\n",
       "      <td>Pyrimidine metabolism PAO1</td>\n",
       "      <td>3.196342e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Purine metabolism PAO1</td>\n",
       "      <td>Inosine monophosphate biosynthesis, PRPP + glu...</td>\n",
       "      <td>4.428380e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Biotin metabolism PAO1</td>\n",
       "      <td>Fatty acid metabolism PAO1</td>\n",
       "      <td>5.471826e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Pantothenate and CoA biosynthesis PAO1</td>\n",
       "      <td>Valine/isoleucine biosynthesis, pyruvate =&gt; va...</td>\n",
       "      <td>5.844695e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Lysine degradation PAO1</td>\n",
       "      <td>Butanoate metabolism PAO1</td>\n",
       "      <td>5.844695e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>DNA replication PAO1</td>\n",
       "      <td>DNA polymerase III complex, bacteria</td>\n",
       "      <td>8.769749e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>Vitamin B6 metabolism PAO1</td>\n",
       "      <td>Pyridoxal biosynthesis, erythrose-4P =&gt; pyrido...</td>\n",
       "      <td>1.052619e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Mismatch repair PAO1</td>\n",
       "      <td>DNA polymerase III complex, bacteria</td>\n",
       "      <td>3.202582e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Fatty acid degradation PAO1</td>\n",
       "      <td>beta-Alanine metabolism PAO1</td>\n",
       "      <td>3.935127e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Bacterial secretion system PAO1</td>\n",
       "      <td>Type III secretion system</td>\n",
       "      <td>3.998219e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>Synthesis and degradation of ketone bodies PAO1</td>\n",
       "      <td>Butanoate metabolism PAO1</td>\n",
       "      <td>4.094787e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>Valine, leucine and isoleucine degradation PAO1</td>\n",
       "      <td>Lysine degradation PAO1</td>\n",
       "      <td>4.094787e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>Pyruvate metabolism PAO1</td>\n",
       "      <td>Propanoate metabolism PAO1</td>\n",
       "      <td>4.324502e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>Benzoate degradation PAO1</td>\n",
       "      <td>Lysine degradation PAO1</td>\n",
       "      <td>5.940569e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>2-Oxocarboxylic acid metabolism PAO1</td>\n",
       "      <td>Valine/isoleucine biosynthesis, pyruvate =&gt; va...</td>\n",
       "      <td>8.453133e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Riboflavin metabolism PAO1</td>\n",
       "      <td>Riboflavin biosynthesis, GTP =&gt; riboflavin/FMN...</td>\n",
       "      <td>8.453133e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>Phenylalanine, tyrosine and tryptophan biosynt...</td>\n",
       "      <td>Tryptophan biosynthesis, chorismate =&gt; tryptophan</td>\n",
       "      <td>8.453133e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>Homologous recombination PAO1</td>\n",
       "      <td>DNA replication PAO1</td>\n",
       "      <td>1.108889e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>Citrate cycle (TCA cycle) PAO1</td>\n",
       "      <td>Pyruvate oxidation, pyruvate =&gt; acetyl-CoA</td>\n",
       "      <td>1.157571e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Assimilatory sulfate reduction, sulfate =&gt; H2S</td>\n",
       "      <td>Sulfate-sulfur assimilation</td>\n",
       "      <td>1.897639e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Glycolysis / Gluconeogenesis PAO1</td>\n",
       "      <td>Pyruvate oxidation, pyruvate =&gt; acetyl-CoA</td>\n",
       "      <td>3.889847e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Glycolysis / Gluconeogenesis PAO1</td>\n",
       "      <td>Gluconeogenesis, oxaloacetate =&gt; fructose-6P</td>\n",
       "      <td>3.889847e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Homologous recombination PAO1</td>\n",
       "      <td>DNA polymerase III complex, bacteria</td>\n",
       "      <td>5.099497e-14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  pw0  \\\n",
       "0                                       Ribosome PAO1   \n",
       "1                          Fatty acid metabolism PAO1   \n",
       "2                      Citrate cycle (TCA cycle) PAO1   \n",
       "3                     Bacterial secretion system PAO1   \n",
       "4   Citrate cycle, second carbon oxidation, 2-oxog...   \n",
       "5     Valine, leucine and isoleucine degradation PAO1   \n",
       "6                              Biotin metabolism PAO1   \n",
       "7                        Fatty acid biosynthesis PAO1   \n",
       "8                         Fatty acid degradation PAO1   \n",
       "9                      Citrate cycle (TCA cycle) PAO1   \n",
       "10                        Fatty acid degradation PAO1   \n",
       "11    Valine, leucine and isoleucine degradation PAO1   \n",
       "12                   Aminoacyl-tRNA biosynthesis PAO1   \n",
       "13                Fatty acid biosynthesis, elongation   \n",
       "14                         Fatty acid metabolism PAO1   \n",
       "15                        Fatty acid degradation PAO1   \n",
       "16                           Lysine biosynthesis PAO1   \n",
       "17   Valine, leucine and isoleucine biosynthesis PAO1   \n",
       "18                             Biotin metabolism PAO1   \n",
       "19                        Phenazine biosynthesis PAO1   \n",
       "20                         Tryptophan metabolism PAO1   \n",
       "21                       Fatty acid biosynthesis PAO1   \n",
       "22                                Quorum sensing PAO1   \n",
       "23                         Fatty acid metabolism PAO1   \n",
       "24   Valine, leucine and isoleucine biosynthesis PAO1   \n",
       "25                         Propanoate metabolism PAO1   \n",
       "26    Valine, leucine and isoleucine degradation PAO1   \n",
       "27                                Protein export PAO1   \n",
       "28                           Folate biosynthesis PAO1   \n",
       "29                             Biotin metabolism PAO1   \n",
       "..                                                ...   \n",
       "45                         Arginine biosynthesis PAO1   \n",
       "46      Imipenem resistance, repression of porin OprD   \n",
       "47                        Fatty acid degradation PAO1   \n",
       "48   Valine, leucine and isoleucine biosynthesis PAO1   \n",
       "49                        Fatty acid degradation PAO1   \n",
       "50                               DNA replication PAO1   \n",
       "51                       beta-Alanine metabolism PAO1   \n",
       "52                             Purine metabolism PAO1   \n",
       "53                             Purine metabolism PAO1   \n",
       "54                             Biotin metabolism PAO1   \n",
       "55             Pantothenate and CoA biosynthesis PAO1   \n",
       "56                            Lysine degradation PAO1   \n",
       "57                               DNA replication PAO1   \n",
       "58                         Vitamin B6 metabolism PAO1   \n",
       "59                               Mismatch repair PAO1   \n",
       "60                        Fatty acid degradation PAO1   \n",
       "61                    Bacterial secretion system PAO1   \n",
       "62    Synthesis and degradation of ketone bodies PAO1   \n",
       "63    Valine, leucine and isoleucine degradation PAO1   \n",
       "64                           Pyruvate metabolism PAO1   \n",
       "65                          Benzoate degradation PAO1   \n",
       "66               2-Oxocarboxylic acid metabolism PAO1   \n",
       "67                         Riboflavin metabolism PAO1   \n",
       "68  Phenylalanine, tyrosine and tryptophan biosynt...   \n",
       "69                      Homologous recombination PAO1   \n",
       "70                     Citrate cycle (TCA cycle) PAO1   \n",
       "71     Assimilatory sulfate reduction, sulfate => H2S   \n",
       "72                  Glycolysis / Gluconeogenesis PAO1   \n",
       "73                  Glycolysis / Gluconeogenesis PAO1   \n",
       "74                      Homologous recombination PAO1   \n",
       "\n",
       "                                                  pw1  adjusted_pval  \n",
       "0                                  Ribosome, bacteria   4.826194e-90  \n",
       "1                        Fatty acid biosynthesis PAO1   5.252269e-41  \n",
       "2              Citrate cycle (TCA cycle, Krebs cycle)   6.427679e-40  \n",
       "3                   Type II general secretion pathway   4.308885e-39  \n",
       "4              Citrate cycle (TCA cycle, Krebs cycle)   2.499061e-33  \n",
       "5                          Propanoate metabolism PAO1   6.428247e-33  \n",
       "6   Pimeloyl-ACP biosynthesis, BioC-BioH pathway, ...   1.409373e-31  \n",
       "7                 Fatty acid biosynthesis, elongation   1.853222e-31  \n",
       "8                          Fatty acid metabolism PAO1   5.967364e-31  \n",
       "9   Citrate cycle, second carbon oxidation, 2-oxog...   1.258273e-29  \n",
       "10    Valine, leucine and isoleucine degradation PAO1   4.631031e-29  \n",
       "11  Leucine degradation, leucine => acetoacetate +...   1.873092e-27  \n",
       "12           Aminoacyl-tRNA biosynthesis, prokaryotes   3.641385e-27  \n",
       "13  Pimeloyl-ACP biosynthesis, BioC-BioH pathway, ...   3.641385e-27  \n",
       "14                Fatty acid biosynthesis, elongation   1.342633e-25  \n",
       "15                                     beta-Oxidation   1.862063e-25  \n",
       "16  Lysine biosynthesis, succinyl-DAP pathway, asp...   2.719932e-25  \n",
       "17  Isoleucine biosynthesis, threonine => 2-oxobut...   3.267995e-25  \n",
       "18                Fatty acid biosynthesis, elongation   2.512307e-24  \n",
       "19  Pyocyanine biosynthesis, chorismate => pyocyanine   4.340715e-24  \n",
       "20                            Lysine degradation PAO1   1.646385e-23  \n",
       "21  Pimeloyl-ACP biosynthesis, BioC-BioH pathway, ...   3.174955e-23  \n",
       "22                        Phenazine biosynthesis PAO1   5.955826e-23  \n",
       "23                                     beta-Oxidation   2.398959e-22  \n",
       "24               2-Oxocarboxylic acid metabolism PAO1   4.864250e-22  \n",
       "25                         Fatty acid metabolism PAO1   1.038392e-21  \n",
       "26                       beta-Alanine metabolism PAO1   1.144403e-21  \n",
       "27                             Sec (secretion) system   1.676715e-21  \n",
       "28          Tetrahydrofolate biosynthesis, GTP => THF   1.822326e-21  \n",
       "29                       Fatty acid biosynthesis PAO1   2.184256e-20  \n",
       "..                                                ...            ...  \n",
       "45   Alanine, aspartate and glutamate metabolism PAO1   4.212918e-17  \n",
       "46  CusS-CusR (copper tolerance) two-component reg...   8.324803e-17  \n",
       "47                         Tryptophan metabolism PAO1   1.059795e-16  \n",
       "48  Valine/isoleucine biosynthesis, pyruvate => va...   1.397409e-16  \n",
       "49                          Benzoate degradation PAO1   1.608228e-16  \n",
       "50                               Mismatch repair PAO1   2.628312e-16  \n",
       "51                                     beta-Oxidation   2.898540e-16  \n",
       "52                         Pyrimidine metabolism PAO1   3.196342e-16  \n",
       "53  Inosine monophosphate biosynthesis, PRPP + glu...   4.428380e-16  \n",
       "54                         Fatty acid metabolism PAO1   5.471826e-16  \n",
       "55  Valine/isoleucine biosynthesis, pyruvate => va...   5.844695e-16  \n",
       "56                          Butanoate metabolism PAO1   5.844695e-16  \n",
       "57               DNA polymerase III complex, bacteria   8.769749e-16  \n",
       "58  Pyridoxal biosynthesis, erythrose-4P => pyrido...   1.052619e-15  \n",
       "59               DNA polymerase III complex, bacteria   3.202582e-15  \n",
       "60                       beta-Alanine metabolism PAO1   3.935127e-15  \n",
       "61                          Type III secretion system   3.998219e-15  \n",
       "62                          Butanoate metabolism PAO1   4.094787e-15  \n",
       "63                            Lysine degradation PAO1   4.094787e-15  \n",
       "64                         Propanoate metabolism PAO1   4.324502e-15  \n",
       "65                            Lysine degradation PAO1   5.940569e-15  \n",
       "66  Valine/isoleucine biosynthesis, pyruvate => va...   8.453133e-15  \n",
       "67  Riboflavin biosynthesis, GTP => riboflavin/FMN...   8.453133e-15  \n",
       "68  Tryptophan biosynthesis, chorismate => tryptophan   8.453133e-15  \n",
       "69                               DNA replication PAO1   1.108889e-14  \n",
       "70         Pyruvate oxidation, pyruvate => acetyl-CoA   1.157571e-14  \n",
       "71                        Sulfate-sulfur assimilation   1.897639e-14  \n",
       "72         Pyruvate oxidation, pyruvate => acetyl-CoA   3.889847e-14  \n",
       "73       Gluconeogenesis, oxaloacetate => fructose-6P   3.889847e-14  \n",
       "74               DNA polymerase III complex, bacteria   5.099497e-14  \n",
       "\n",
       "[75 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "significant_edges = create_overlap_based_network(kegg_pathways_dict, alpha=0.05)\n",
    "print(\"Number of edges in the PAO1 KEGG overlap-based network: {0}\".format(\n",
    "    len(significant_edges)))\n",
    "sorted_significant_edges = sorted(significant_edges, key=lambda tup: tup[1])\n",
    "\n",
    "pathways_in_network = set()\n",
    "edges = []\n",
    "for (pw0, pw1), adjusted_pval in sorted_significant_edges:\n",
    "    row = {\"pw0\": pw0, \"pw1\": pw1, \"adjusted_pval\": adjusted_pval}\n",
    "    pathways_in_network.add(pw0)\n",
    "    pathways_in_network.add(pw1)\n",
    "    edges.append(row)\n",
    "\n",
    "print(\"Number of pathways in overlap-based network: {0}\".format(\n",
    "    len(pathways_in_network)))\n",
    "    \n",
    "show_N_edges = 75\n",
    "edges_df = pd.DataFrame(edges, columns=[\"pw0\", \"pw1\", \"adjusted_pval\"])\n",
    "edges_df.head(show_N_edges)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TCGA PID network summary statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pw0</th>\n",
       "      <th>pw1</th>\n",
       "      <th>weight</th>\n",
       "      <th>features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E2F</td>\n",
       "      <td>PLK1</td>\n",
       "      <td>6.543478</td>\n",
       "      <td>19.0 29.0 39.0 44.0 59.0 114.0 115.0 158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AURORA B</td>\n",
       "      <td>E2F</td>\n",
       "      <td>6.531936</td>\n",
       "      <td>19.0 29.0 39.0 44.0 59.0 114.0 115.0 158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AURORA B</td>\n",
       "      <td>PLK1</td>\n",
       "      <td>6.341347</td>\n",
       "      <td>19.0 29.0 39.0 44.0 59.0 114.0 115.0 158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E2F</td>\n",
       "      <td>FOXM1</td>\n",
       "      <td>5.974458</td>\n",
       "      <td>19.0 29.0 39.0 44.0 59.0 115.0 158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AURORA B</td>\n",
       "      <td>FOXM1</td>\n",
       "      <td>5.873026</td>\n",
       "      <td>19.0 29.0 39.0 44.0 59.0 115.0 158.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        pw0    pw1    weight                                    features\n",
       "0       E2F   PLK1  6.543478  19.0 29.0 39.0 44.0 59.0 114.0 115.0 158.0\n",
       "1  AURORA B    E2F  6.531936  19.0 29.0 39.0 44.0 59.0 114.0 115.0 158.0\n",
       "2  AURORA B   PLK1  6.341347  19.0 29.0 39.0 44.0 59.0 114.0 115.0 158.0\n",
       "3       E2F  FOXM1  5.974458        19.0 29.0 39.0 44.0 59.0 115.0 158.0\n",
       "4  AURORA B  FOXM1  5.873026        19.0 29.0 39.0 44.0 59.0 115.0 158.0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcga_pid_network = pd.read_table(tcga_pid_edges_file)\n",
    "tcga_pid_network.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# shorten the PID pathway names so that they match the ones\n",
    "# we use in the publication\n",
    "def shorten_tcga_pid(pathway_name):\n",
    "    REMOVE_SUFFIX = \"PATHWAY\"\n",
    "    pathway_short = None\n",
    "    split_on_underscores = pathway_name.split(\"_\")\n",
    "    if split_on_underscores[-1] == REMOVE_SUFFIX:\n",
    "        pathway_short = \" \".join(split_on_underscores[:-1])\n",
    "    else:\n",
    "        pathway_short = \" \".join(split_on_underscores)\n",
    "    return pathway_short\n",
    "\n",
    "pid_pathways_dict = load_pathway_definitions_file(\n",
    "    pid_pathways, shorten_tcga_pid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the TCGA PID pathway-pathway co-occurrence network:\n",
      "Number of distinct pathways in the network: 57\n",
      "Number of edges in the network: 119\n",
      "Number of edges in the network where the 2 pathways share genes: 96 (average similarity by Jaccard Index: 0.05783)\n"
     ]
    }
   ],
   "source": [
    "pathways_in_pid_network = set(\n",
    "    tcga_pid_network[\"pw0\"].tolist() + tcga_pid_network[\"pw1\"].tolist())\n",
    "\n",
    "n_overlapping = 0\n",
    "edge_jaccards = []\n",
    "for index, row in tcga_pid_network.iterrows():\n",
    "    pw0_genes = pid_pathways_dict[row[\"pw0\"]] \n",
    "    pw1_genes = pid_pathways_dict[row[\"pw1\"]]\n",
    "    overlap = pw0_genes & pw1_genes\n",
    "    if overlap:\n",
    "        n_overlapping += 1\n",
    "    jaccard = float(len(overlap)) / (len(pw0_genes) + len(pw1_genes) - len(overlap))\n",
    "    edge_jaccards.append(jaccard)\n",
    "\n",
    "print(\"For the TCGA PID pathway-pathway co-occurrence network:\")\n",
    "print(\"Number of distinct pathways in the network: {0}\".format(\n",
    "    len(pathways_in_pid_network)))\n",
    "print(\"Number of edges in the network: {0}\".format(\n",
    "    tcga_pid_network.shape[0]))\n",
    "print(\"Number of edges in the network where the 2 pathways \"\n",
    "      \"share genes: {0} (average similarity by Jaccard Index: {1:.4})\".format(\n",
    "          n_overlapping, np.average(edge_jaccards)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison to a network based only on overlapping genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just as we did for the PAO1 KEGG case, we construct a network only from gene overlap (FDR-corrected hypergeometric test < 0.05) for the TCGA PID case. \n",
    "The PathCORE-T network had 119 edges between 57 pathways.\n",
    "The overlap-based network (see below) had 3826 edges between 196 pathways. \n",
    "The fact that the overlap-based network is so dense suggests that there is a high degree of overlap between PID pathways.\n",
    "96 of the edges in the PathCORE-T network were between pathways that shared genes, but the average Jaccard Index for these pathway pairs was only 0.058. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of edges in the PID overlap-based network: 3826\n",
      "Number of pathways in overlap-based network: 196\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pw0</th>\n",
       "      <th>pw1</th>\n",
       "      <th>adjusted_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCR</td>\n",
       "      <td>CD8 TCR</td>\n",
       "      <td>2.070913e-76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERBB1 DOWNSTREAM</td>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>1.431885e-53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IL8 CXCR2</td>\n",
       "      <td>IL8 CXCR1</td>\n",
       "      <td>2.000459e-42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AVB3 INTEGRIN</td>\n",
       "      <td>SYNDECAN 1</td>\n",
       "      <td>4.320553e-38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENDOTHELIN</td>\n",
       "      <td>LYSOPHOSPHOLIPID</td>\n",
       "      <td>4.192069e-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CDC42</td>\n",
       "      <td>RAC1</td>\n",
       "      <td>4.192069e-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GMCSF</td>\n",
       "      <td>IL2 1PATHWAY</td>\n",
       "      <td>3.790383e-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FCER1</td>\n",
       "      <td>BCR 5PATHWAY</td>\n",
       "      <td>5.974713e-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GMCSF</td>\n",
       "      <td>IL3</td>\n",
       "      <td>5.974713e-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDC42</td>\n",
       "      <td>ERBB1 DOWNSTREAM</td>\n",
       "      <td>5.974713e-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>INTEGRIN1</td>\n",
       "      <td>INTEGRIN3</td>\n",
       "      <td>7.590876e-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>S1P S1P3</td>\n",
       "      <td>S1P S1P2</td>\n",
       "      <td>4.782879e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NFAT TFPATHWAY</td>\n",
       "      <td>TCR CALCIUM</td>\n",
       "      <td>4.782879e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ERBB1 DOWNSTREAM</td>\n",
       "      <td>ERBB2 ERBB3</td>\n",
       "      <td>6.337655e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ERBB1 RECEPTOR PROXIMAL</td>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>9.436802e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SHP2</td>\n",
       "      <td>IL2 1PATHWAY</td>\n",
       "      <td>3.021094e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>KIT</td>\n",
       "      <td>EPO</td>\n",
       "      <td>3.176346e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ER NONGENOMIC</td>\n",
       "      <td>THROMBIN PAR1</td>\n",
       "      <td>4.363902e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>S1P S1P3</td>\n",
       "      <td>S1P S1P1</td>\n",
       "      <td>4.408738e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TXA2PATHWAY</td>\n",
       "      <td>IL8 CXCR1</td>\n",
       "      <td>8.932892e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>IL2 PI3K</td>\n",
       "      <td>IL2 STAT5</td>\n",
       "      <td>1.063465e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>TRAIL</td>\n",
       "      <td>FAS</td>\n",
       "      <td>2.236480e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>LYSOPHOSPHOLIPID</td>\n",
       "      <td>ER NONGENOMIC</td>\n",
       "      <td>3.703095e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENDOTHELIN</td>\n",
       "      <td>S1P S1P2</td>\n",
       "      <td>8.081708e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>IL12 2PATHWAY</td>\n",
       "      <td>IL12 STAT4</td>\n",
       "      <td>9.125227e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>MET</td>\n",
       "      <td>FAK</td>\n",
       "      <td>1.355788e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>THROMBIN PAR4</td>\n",
       "      <td>THROMBIN PAR1</td>\n",
       "      <td>1.460504e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ERBB2 ERBB3</td>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>1.460504e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>FRA</td>\n",
       "      <td>AP1</td>\n",
       "      <td>3.839984e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HIV NEF</td>\n",
       "      <td>CASPASE</td>\n",
       "      <td>3.911065e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>MET</td>\n",
       "      <td>RET</td>\n",
       "      <td>2.760806e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>IL27</td>\n",
       "      <td>IL23</td>\n",
       "      <td>2.888862e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>ERBB1 DOWNSTREAM</td>\n",
       "      <td>PI3K PLC TRK</td>\n",
       "      <td>3.243767e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>INSULIN</td>\n",
       "      <td>MET</td>\n",
       "      <td>3.516756e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>CONE</td>\n",
       "      <td>RHODOPSIN</td>\n",
       "      <td>3.516756e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>MET</td>\n",
       "      <td>IGF1</td>\n",
       "      <td>4.361016e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>ERBB1 RECEPTOR PROXIMAL</td>\n",
       "      <td>ERBB2 ERBB3</td>\n",
       "      <td>4.361016e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>SHP2</td>\n",
       "      <td>TRKR</td>\n",
       "      <td>5.513524e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>TXA2PATHWAY</td>\n",
       "      <td>IL8 CXCR2</td>\n",
       "      <td>6.168136e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>ER NONGENOMIC</td>\n",
       "      <td>AR NONGENOMIC</td>\n",
       "      <td>6.339960e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>IL12 2PATHWAY</td>\n",
       "      <td>IL23</td>\n",
       "      <td>6.622137e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>INSULIN</td>\n",
       "      <td>IGF1</td>\n",
       "      <td>1.835380e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>IL2 1PATHWAY</td>\n",
       "      <td>ERBB2 ERBB3</td>\n",
       "      <td>2.239092e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>CERAMIDE</td>\n",
       "      <td>HIV NEF</td>\n",
       "      <td>2.297568e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>FCER1</td>\n",
       "      <td>MET</td>\n",
       "      <td>2.615614e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>ENDOTHELIN</td>\n",
       "      <td>S1P S1P3</td>\n",
       "      <td>6.077232e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>ERBB1 RECEPTOR PROXIMAL</td>\n",
       "      <td>ERBB1 INTERNALIZATION</td>\n",
       "      <td>7.317938e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>RET</td>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>1.245196e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>S1P META</td>\n",
       "      <td>S1P S1P2</td>\n",
       "      <td>1.469832e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>NETRIN</td>\n",
       "      <td>FAK</td>\n",
       "      <td>1.654028e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>TCPTP</td>\n",
       "      <td>ERBB1 RECEPTOR PROXIMAL</td>\n",
       "      <td>1.743474e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>BETA CATENIN DEG</td>\n",
       "      <td>WNT CANONICAL</td>\n",
       "      <td>2.032783e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>TCR RAS</td>\n",
       "      <td>CD8 TCR DOWNSTREAM</td>\n",
       "      <td>2.089632e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>FCER1</td>\n",
       "      <td>GMCSF</td>\n",
       "      <td>2.128036e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>ENDOTHELIN</td>\n",
       "      <td>THROMBIN PAR1</td>\n",
       "      <td>2.170068e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>PI3KCI</td>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>2.352243e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>PDGFRB</td>\n",
       "      <td>RAC1</td>\n",
       "      <td>2.430969e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>RET</td>\n",
       "      <td>FAK</td>\n",
       "      <td>2.462213e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>ERBB2 ERBB3</td>\n",
       "      <td>CXCR3</td>\n",
       "      <td>2.604203e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>EPHB FWD</td>\n",
       "      <td>RET</td>\n",
       "      <td>3.328702e-18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        pw0                      pw1  adjusted_pval\n",
       "0                       TCR                  CD8 TCR   2.070913e-76\n",
       "1          ERBB1 DOWNSTREAM                   PDGFRB   1.431885e-53\n",
       "2                 IL8 CXCR2                IL8 CXCR1   2.000459e-42\n",
       "3             AVB3 INTEGRIN               SYNDECAN 1   4.320553e-38\n",
       "4                ENDOTHELIN         LYSOPHOSPHOLIPID   4.192069e-34\n",
       "5                     CDC42                     RAC1   4.192069e-34\n",
       "6                     GMCSF             IL2 1PATHWAY   3.790383e-26\n",
       "7                     FCER1             BCR 5PATHWAY   5.974713e-26\n",
       "8                     GMCSF                      IL3   5.974713e-26\n",
       "9                     CDC42         ERBB1 DOWNSTREAM   5.974713e-26\n",
       "10                INTEGRIN1                INTEGRIN3   7.590876e-26\n",
       "11                 S1P S1P3                 S1P S1P2   4.782879e-25\n",
       "12           NFAT TFPATHWAY              TCR CALCIUM   4.782879e-25\n",
       "13         ERBB1 DOWNSTREAM              ERBB2 ERBB3   6.337655e-25\n",
       "14  ERBB1 RECEPTOR PROXIMAL                   PDGFRB   9.436802e-25\n",
       "15                     SHP2             IL2 1PATHWAY   3.021094e-24\n",
       "16                      KIT                      EPO   3.176346e-24\n",
       "17            ER NONGENOMIC            THROMBIN PAR1   4.363902e-24\n",
       "18                 S1P S1P3                 S1P S1P1   4.408738e-24\n",
       "19              TXA2PATHWAY                IL8 CXCR1   8.932892e-24\n",
       "20                 IL2 PI3K                IL2 STAT5   1.063465e-23\n",
       "21                    TRAIL                      FAS   2.236480e-23\n",
       "22         LYSOPHOSPHOLIPID            ER NONGENOMIC   3.703095e-23\n",
       "23               ENDOTHELIN                 S1P S1P2   8.081708e-23\n",
       "24            IL12 2PATHWAY               IL12 STAT4   9.125227e-23\n",
       "25                      MET                      FAK   1.355788e-22\n",
       "26            THROMBIN PAR4            THROMBIN PAR1   1.460504e-22\n",
       "27              ERBB2 ERBB3                   PDGFRB   1.460504e-22\n",
       "28                      FRA                      AP1   3.839984e-22\n",
       "29                  HIV NEF                  CASPASE   3.911065e-22\n",
       "..                      ...                      ...            ...\n",
       "45                      MET                      RET   2.760806e-20\n",
       "46                     IL27                     IL23   2.888862e-20\n",
       "47         ERBB1 DOWNSTREAM             PI3K PLC TRK   3.243767e-20\n",
       "48                  INSULIN                      MET   3.516756e-20\n",
       "49                     CONE                RHODOPSIN   3.516756e-20\n",
       "50                      MET                     IGF1   4.361016e-20\n",
       "51  ERBB1 RECEPTOR PROXIMAL              ERBB2 ERBB3   4.361016e-20\n",
       "52                     SHP2                     TRKR   5.513524e-20\n",
       "53              TXA2PATHWAY                IL8 CXCR2   6.168136e-20\n",
       "54            ER NONGENOMIC            AR NONGENOMIC   6.339960e-20\n",
       "55            IL12 2PATHWAY                     IL23   6.622137e-20\n",
       "56                  INSULIN                     IGF1   1.835380e-19\n",
       "57             IL2 1PATHWAY              ERBB2 ERBB3   2.239092e-19\n",
       "58                 CERAMIDE                  HIV NEF   2.297568e-19\n",
       "59                    FCER1                      MET   2.615614e-19\n",
       "60               ENDOTHELIN                 S1P S1P3   6.077232e-19\n",
       "61  ERBB1 RECEPTOR PROXIMAL    ERBB1 INTERNALIZATION   7.317938e-19\n",
       "62                      RET                   PDGFRB   1.245196e-18\n",
       "63                 S1P META                 S1P S1P2   1.469832e-18\n",
       "64                   NETRIN                      FAK   1.654028e-18\n",
       "65                    TCPTP  ERBB1 RECEPTOR PROXIMAL   1.743474e-18\n",
       "66         BETA CATENIN DEG            WNT CANONICAL   2.032783e-18\n",
       "67                  TCR RAS       CD8 TCR DOWNSTREAM   2.089632e-18\n",
       "68                    FCER1                    GMCSF   2.128036e-18\n",
       "69               ENDOTHELIN            THROMBIN PAR1   2.170068e-18\n",
       "70                   PI3KCI                   PDGFRB   2.352243e-18\n",
       "71                   PDGFRB                     RAC1   2.430969e-18\n",
       "72                      RET                      FAK   2.462213e-18\n",
       "73              ERBB2 ERBB3                    CXCR3   2.604203e-18\n",
       "74                 EPHB FWD                      RET   3.328702e-18\n",
       "\n",
       "[75 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "significant_edges = create_overlap_based_network(pid_pathways_dict, alpha=0.05)\n",
    "print(\"Number of edges in the PID overlap-based network: {0}\".format(\n",
    "    len(significant_edges)))\n",
    "sorted_significant_edges = sorted(significant_edges, key=lambda tup: tup[1])\n",
    "\n",
    "pathways_in_network = set()\n",
    "edges = []\n",
    "for (pw0, pw1), adjusted_pval in sorted_significant_edges:\n",
    "    row = {\"pw0\": pw0, \"pw1\": pw1, \"adjusted_pval\": adjusted_pval}\n",
    "    pathways_in_network.add(pw0)\n",
    "    pathways_in_network.add(pw1)\n",
    "    edges.append(row)\n",
    "\n",
    "print(\"Number of pathways in overlap-based network: {0}\".format(\n",
    "    len(pathways_in_network)))\n",
    "    \n",
    "show_N_edges = 75\n",
    "edges_df = pd.DataFrame(edges, columns=[\"pw0\", \"pw1\", \"adjusted_pval\"])\n",
    "edges_df.head(show_N_edges)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
