{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Additional information about the PAO1 and TCGA pathway-pathway networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import fisher_exact\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "data_dir = os.path.join(\"..\", \"data\")\n",
    "pao1_data = os.path.join(data_dir, \"pao1_data\")\n",
    "tcga_data = os.path.join(data_dir, \"tcga_data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "pao1_kegg_edges_file = os.path.join(\n",
    "    pao1_data, \"eADAGE_analysis\", \"permutation_test_n=10000\", \"filtered_network.tsv\")\n",
    "tcga_pid_edges_file = os.path.join(\n",
    "    tcga_data, \"NMF_analysis\", \"permutation_test_n=10000\", \"filtered_network.tsv\")\n",
    "\n",
    "kegg_pathways = os.path.join(pao1_data, \"pseudomonas_KEGG_terms.txt\")\n",
    "pid_pathways = os.path.join(tcga_data, \"PID_pathway_definitions.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def load_pathway_definitions_file(pathways_file, shorten_pathway_name):\n",
    "    pathway_definitions_df = pd.read_table(\n",
    "        pathways_file, header=None, names=[\"pw\", \"size\", \"genes\"])\n",
    "    pathway_definitions_df[\"genes\"] = pathway_definitions_df[\"genes\"].map(\n",
    "        lambda x: x.split(\";\"))\n",
    "    pathway_definitions_df.set_index(\"pw\", inplace=True)\n",
    "\n",
    "    pathway_definitions_map = {}\n",
    "    for index, row in pathway_definitions_df.iterrows():\n",
    "        pathway = shorten_pathway_name(index)\n",
    "        pathway_definitions_map[pathway] = set(row[\"genes\"])\n",
    "    return pathway_definitions_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pw0</th>\n",
       "      <th>pw1</th>\n",
       "      <th>weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ribosome PAO1</td>\n",
       "      <td>Ribosome, bacteria</td>\n",
       "      <td>26.089640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Phosphonate and phosphinate metabolism PAO1</td>\n",
       "      <td>Phosphonate transport system</td>\n",
       "      <td>25.455394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Macrolide resistance, MacAB-TolC transporter</td>\n",
       "      <td>Zinc transport system</td>\n",
       "      <td>23.170151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Phosphonate and phosphinate metabolism PAO1</td>\n",
       "      <td>Type II general secretion pathway</td>\n",
       "      <td>21.168310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Glycine betaine/proline transport system</td>\n",
       "      <td>Glycine, serine and threonine metabolism PAO1</td>\n",
       "      <td>19.221697</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            pw0  \\\n",
       "0                                 Ribosome PAO1   \n",
       "1   Phosphonate and phosphinate metabolism PAO1   \n",
       "2  Macrolide resistance, MacAB-TolC transporter   \n",
       "3   Phosphonate and phosphinate metabolism PAO1   \n",
       "4      Glycine betaine/proline transport system   \n",
       "\n",
       "                                             pw1     weight  \n",
       "0                             Ribosome, bacteria  26.089640  \n",
       "1                   Phosphonate transport system  25.455394  \n",
       "2                          Zinc transport system  23.170151  \n",
       "3              Type II general secretion pathway  21.168310  \n",
       "4  Glycine, serine and threonine metabolism PAO1  19.221697  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pao1_kegg_network = pd.read_table(pao1_kegg_edges_file)\n",
    "pao1_kegg_network.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# shorten the PAO1 KEGG pathway names so that they match the ones\n",
    "# we use in the publication\n",
    "def shorten_pao1_kegg(pathway_name):\n",
    "    REMOVE_SUFFIX = \"- Pseudomonas aeruginosa PAO1\"\n",
    "    pathway_short = None\n",
    "    split_label = pathway_name.split(\" \", 1)\n",
    "    if len(split_label) > 1:\n",
    "        pathway_short = split_label[1]\n",
    "    else:\n",
    "        pathway_short = split_label[0]\n",
    "    if REMOVE_SUFFIX in pathway_short:\n",
    "        remove_from_index = pathway_short.index(REMOVE_SUFFIX)\n",
    "        return \"{0}PAO1\".format(pathway_short[:remove_from_index])\n",
    "    return pathway_short.strip()\n",
    "\n",
    "kegg_pathways_dict = load_pathway_definitions_file(\n",
    "    kegg_pathways, shorten_pao1_kegg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PAO1 KEGG network summary statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the PAO1 KEGG pathway-pathway co-occurrence network:\n",
      "Number of distinct pathways in the network: 89\n",
      "Number of edges in the network: 203\n",
      "Number of edges in the network where the 2 pathways share genes: 35 (average similarity by Jaccard Index: 0.03482)\n"
     ]
    }
   ],
   "source": [
    "pathways_in_kegg_network = set(\n",
    "    pao1_kegg_network[\"pw0\"].tolist() + pao1_kegg_network[\"pw1\"].tolist())\n",
    "\n",
    "n_overlapping = 0\n",
    "edge_jaccards = []\n",
    "for index, row in pao1_kegg_network.iterrows():\n",
    "    pw0_genes = kegg_pathways_dict[row[\"pw0\"]] \n",
    "    pw1_genes = kegg_pathways_dict[row[\"pw1\"]]\n",
    "    overlap = pw0_genes & pw1_genes\n",
    "    if overlap:\n",
    "        n_overlapping += 1\n",
    "    jaccard = float(len(overlap)) / (len(pw0_genes) + len(pw1_genes) - len(overlap))\n",
    "    edge_jaccards.append(jaccard)        \n",
    "        \n",
    "print(\"For the PAO1 KEGG pathway-pathway co-occurrence network:\")\n",
    "print(\"Number of distinct pathways in the network: {0}\".format(\n",
    "    len(pathways_in_kegg_network)))\n",
    "print(\"Number of edges in the network: {0}\".format(\n",
    "    pao1_kegg_network.shape[0]))\n",
    "print(\"Number of edges in the network where the 2 pathways \"\n",
    "      \"share genes: {0} (average similarity by Jaccard Index: {1:.4})\".format(\n",
    "          n_overlapping, np.average(edge_jaccards)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison to a network based only on overlapping genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_overlap_based_network(pathway_genes_map, alpha=0.05):\n",
    "    all_genes = set().union(*pathway_genes_map.values())\n",
    "    n_genes = len(all_genes)\n",
    "    \n",
    "    pvalues_list = []\n",
    "    pathway_pairs = list(itertools.combinations(\n",
    "        pathway_genes_map.keys(), 2))\n",
    "    for pw0, pw1 in pathway_pairs:\n",
    "        pw0_genes = pathway_genes_map[pw0]\n",
    "        pw1_genes = pathway_genes_map[pw1]\n",
    "        n_common = len(pw0_genes & pw1_genes)\n",
    "        n_pw0 = len(pw0_genes - pw1_genes)\n",
    "        n_pw1 = len(pw1_genes - pw0_genes)\n",
    "        n_all_other_genes = n_genes - n_pw0 - n_pw1 - n_common\n",
    "        contingency_table = np.array(\n",
    "            [[n_common, n_pw0],\n",
    "             [n_pw1, n_all_other_genes]])\n",
    "        _, pvalue = fisher_exact(\n",
    "            contingency_table, alternative=\"greater\")\n",
    "        pvalues_list.append(pvalue)\n",
    "    \n",
    "    below_alpha, qvalues, _, _ = multipletests(\n",
    "        pvalues_list, alpha=alpha, method=\"fdr_bh\")\n",
    "    \n",
    "    significant_edges = [(pathway_pairs[ix], qvalues[ix]) for ix, is_below \n",
    "                         in enumerate(below_alpha) if is_below]\n",
    "    return significant_edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of edges in the PAO1 KEGG overlap-based network: 406\n",
      "\n",
      "Pathway\t\tAdjusted p-value\n",
      "('Ribosome PAO1', 'Ribosome, bacteria'):\t0.0000\n",
      "('Fatty acid metabolism PAO1', 'Fatty acid biosynthesis PAO1'):\t0.0000\n",
      "('Citrate cycle (TCA cycle) PAO1', 'Citrate cycle (TCA cycle, Krebs cycle)'):\t0.0000\n",
      "('Bacterial secretion system PAO1', 'Type II general secretion pathway'):\t0.0000\n",
      "('Citrate cycle, second carbon oxidation, 2-oxoglutarate => oxaloacetate', 'Citrate cycle (TCA cycle, Krebs cycle)'):\t0.0000\n",
      "('Valine, leucine and isoleucine degradation PAO1', 'Propanoate metabolism PAO1'):\t0.0000\n",
      "('Biotin metabolism PAO1', 'Pimeloyl-ACP biosynthesis, BioC-BioH pathway, malonyl-ACP => pimeloyl-ACP'):\t0.0000\n",
      "('Fatty acid biosynthesis PAO1', 'Fatty acid biosynthesis, elongation'):\t0.0000\n",
      "('Fatty acid degradation PAO1', 'Fatty acid metabolism PAO1'):\t0.0000\n",
      "('Citrate cycle (TCA cycle) PAO1', 'Citrate cycle, second carbon oxidation, 2-oxoglutarate => oxaloacetate'):\t0.0000\n",
      "('Fatty acid degradation PAO1', 'Valine, leucine and isoleucine degradation PAO1'):\t0.0000\n",
      "('Valine, leucine and isoleucine degradation PAO1', 'Leucine degradation, leucine => acetoacetate + acetyl-CoA'):\t0.0000\n",
      "('Aminoacyl-tRNA biosynthesis PAO1', 'Aminoacyl-tRNA biosynthesis, prokaryotes'):\t0.0000\n",
      "('Fatty acid biosynthesis, elongation', 'Pimeloyl-ACP biosynthesis, BioC-BioH pathway, malonyl-ACP => pimeloyl-ACP'):\t0.0000\n",
      "('Fatty acid metabolism PAO1', 'Fatty acid biosynthesis, elongation'):\t0.0000\n",
      "('Fatty acid degradation PAO1', 'beta-Oxidation'):\t0.0000\n",
      "('Lysine biosynthesis PAO1', 'Lysine biosynthesis, succinyl-DAP pathway, aspartate => lysine'):\t0.0000\n",
      "('Valine, leucine and isoleucine biosynthesis PAO1', 'Isoleucine biosynthesis, threonine => 2-oxobutanoate => isoleucine'):\t0.0000\n",
      "('Biotin metabolism PAO1', 'Fatty acid biosynthesis, elongation'):\t0.0000\n",
      "('Phenazine biosynthesis PAO1', 'Pyocyanine biosynthesis, chorismate => pyocyanine'):\t0.0000\n",
      "('Tryptophan metabolism PAO1', 'Lysine degradation PAO1'):\t0.0000\n",
      "('Fatty acid biosynthesis PAO1', 'Pimeloyl-ACP biosynthesis, BioC-BioH pathway, malonyl-ACP => pimeloyl-ACP'):\t0.0000\n",
      "('Quorum sensing PAO1', 'Phenazine biosynthesis PAO1'):\t0.0000\n",
      "('Fatty acid metabolism PAO1', 'beta-Oxidation'):\t0.0000\n",
      "('Valine, leucine and isoleucine biosynthesis PAO1', '2-Oxocarboxylic acid metabolism PAO1'):\t0.0000\n",
      "('Propanoate metabolism PAO1', 'Fatty acid metabolism PAO1'):\t0.0000\n",
      "('Valine, leucine and isoleucine degradation PAO1', 'beta-Alanine metabolism PAO1'):\t0.0000\n",
      "('Protein export PAO1', 'Sec (secretion) system'):\t0.0000\n",
      "('Folate biosynthesis PAO1', 'Tetrahydrofolate biosynthesis, GTP => THF'):\t0.0000\n",
      "('Biotin metabolism PAO1', 'Fatty acid biosynthesis PAO1'):\t0.0000\n",
      "('Benzoate degradation PAO1', 'Degradation of aromatic compounds PAO1'):\t0.0000\n",
      "('Valine, leucine and isoleucine degradation PAO1', 'Fatty acid metabolism PAO1'):\t0.0000\n",
      "('Oxidative phosphorylation PAO1', 'NADH:quinone oxidoreductase, prokaryotes'):\t0.0000\n",
      "('Valine, leucine and isoleucine degradation PAO1', 'beta-Oxidation'):\t0.0000\n",
      "('Fatty acid degradation PAO1', 'Lysine degradation PAO1'):\t0.0000\n",
      "('Bacterial secretion system PAO1', 'Type VI secretion system'):\t0.0000\n",
      "('Valine/isoleucine biosynthesis, pyruvate => valine / 2-oxobutanoate => isoleucine', 'Isoleucine biosynthesis, threonine => 2-oxobutanoate => isoleucine'):\t0.0000\n",
      "('Cationic antimicrobial peptide (CAMP) resistance PAO1', 'Cationic antimicrobial peptide (CAMP) resistance, arnBCADTEF operon'):\t0.0000\n",
      "('Fatty acid metabolism PAO1', 'Pimeloyl-ACP biosynthesis, BioC-BioH pathway, malonyl-ACP => pimeloyl-ACP'):\t0.0000\n",
      "('Pyruvate metabolism PAO1', 'Glycolysis / Gluconeogenesis PAO1'):\t0.0000\n",
      "('Propanoate metabolism PAO1', 'beta-Alanine metabolism PAO1'):\t0.0000\n",
      "('Citrate cycle (TCA cycle) PAO1', 'Pyruvate metabolism PAO1'):\t0.0000\n",
      "('Fatty acid degradation PAO1', 'Propanoate metabolism PAO1'):\t0.0000\n",
      "('Sulfur metabolism PAO1', 'Sulfate-sulfur assimilation'):\t0.0000\n",
      "('Sulfur metabolism PAO1', 'Sulfonate transport system'):\t0.0000\n",
      "('Arginine biosynthesis PAO1', 'Alanine, aspartate and glutamate metabolism PAO1'):\t0.0000\n",
      "('Imipenem resistance, repression of porin OprD', 'CusS-CusR (copper tolerance) two-component regulatory system'):\t0.0000\n",
      "('Fatty acid degradation PAO1', 'Tryptophan metabolism PAO1'):\t0.0000\n",
      "('Valine, leucine and isoleucine biosynthesis PAO1', 'Valine/isoleucine biosynthesis, pyruvate => valine / 2-oxobutanoate => isoleucine'):\t0.0000\n",
      "('Fatty acid degradation PAO1', 'Benzoate degradation PAO1'):\t0.0000\n",
      "('DNA replication PAO1', 'Mismatch repair PAO1'):\t0.0000\n",
      "('beta-Alanine metabolism PAO1', 'beta-Oxidation'):\t0.0000\n",
      "('Purine metabolism PAO1', 'Pyrimidine metabolism PAO1'):\t0.0000\n",
      "('Purine metabolism PAO1', 'Inosine monophosphate biosynthesis, PRPP + glutamine => IMP'):\t0.0000\n",
      "('Biotin metabolism PAO1', 'Fatty acid metabolism PAO1'):\t0.0000\n",
      "('Pantothenate and CoA biosynthesis PAO1', 'Valine/isoleucine biosynthesis, pyruvate => valine / 2-oxobutanoate => isoleucine'):\t0.0000\n",
      "('Lysine degradation PAO1', 'Butanoate metabolism PAO1'):\t0.0000\n",
      "('DNA replication PAO1', 'DNA polymerase III complex, bacteria'):\t0.0000\n",
      "('Vitamin B6 metabolism PAO1', 'Pyridoxal biosynthesis, erythrose-4P => pyridoxal-5P'):\t0.0000\n",
      "('Mismatch repair PAO1', 'DNA polymerase III complex, bacteria'):\t0.0000\n",
      "('Fatty acid degradation PAO1', 'beta-Alanine metabolism PAO1'):\t0.0000\n",
      "('Bacterial secretion system PAO1', 'Type III secretion system'):\t0.0000\n",
      "('Synthesis and degradation of ketone bodies PAO1', 'Butanoate metabolism PAO1'):\t0.0000\n",
      "('Valine, leucine and isoleucine degradation PAO1', 'Lysine degradation PAO1'):\t0.0000\n",
      "('Pyruvate metabolism PAO1', 'Propanoate metabolism PAO1'):\t0.0000\n",
      "('Benzoate degradation PAO1', 'Lysine degradation PAO1'):\t0.0000\n",
      "('2-Oxocarboxylic acid metabolism PAO1', 'Valine/isoleucine biosynthesis, pyruvate => valine / 2-oxobutanoate => isoleucine'):\t0.0000\n",
      "('Riboflavin metabolism PAO1', 'Riboflavin biosynthesis, GTP => riboflavin/FMN/FAD'):\t0.0000\n",
      "('Phenylalanine, tyrosine and tryptophan biosynthesis PAO1', 'Tryptophan biosynthesis, chorismate => tryptophan'):\t0.0000\n",
      "('Homologous recombination PAO1', 'DNA replication PAO1'):\t0.0000\n",
      "('Citrate cycle (TCA cycle) PAO1', 'Pyruvate oxidation, pyruvate => acetyl-CoA'):\t0.0000\n",
      "('Assimilatory sulfate reduction, sulfate => H2S', 'Sulfate-sulfur assimilation'):\t0.0000\n",
      "('Glycolysis / Gluconeogenesis PAO1', 'Pyruvate oxidation, pyruvate => acetyl-CoA'):\t0.0000\n",
      "('Glycolysis / Gluconeogenesis PAO1', 'Gluconeogenesis, oxaloacetate => fructose-6P'):\t0.0000\n",
      "('Homologous recombination PAO1', 'DNA polymerase III complex, bacteria'):\t0.0000\n",
      "('Nitrogen metabolism PAO1', 'Denitrification, nitrate => nitrogen'):\t0.0000\n",
      "('Homologous recombination PAO1', 'Mismatch repair PAO1'):\t0.0000\n",
      "('Valine, leucine and isoleucine degradation PAO1', 'Butanoate metabolism PAO1'):\t0.0000\n",
      "('Putative spermidine/putrescine transport system', 'Quorum sensing PAO1'):\t0.0000\n",
      "('Propanoate metabolism PAO1', 'Leucine degradation, leucine => acetoacetate + acetyl-CoA'):\t0.0000\n",
      "('Fatty acid biosynthesis PAO1', 'Fatty acid biosynthesis, initiation'):\t0.0000\n",
      "('C5-Branched dibasic acid metabolism PAO1', 'Valine, leucine and isoleucine biosynthesis PAO1'):\t0.0000\n",
      "('Bacterial secretion system PAO1', 'Protein export PAO1'):\t0.0000\n",
      "('Pantothenate and CoA biosynthesis PAO1', 'Isoleucine biosynthesis, threonine => 2-oxobutanoate => isoleucine'):\t0.0000\n",
      "('Porphyrin and chlorophyll metabolism PAO1', 'Cobalamin biosynthesis, cobinamide => cobalamin'):\t0.0000\n",
      "('Propanoate metabolism PAO1', 'beta-Oxidation'):\t0.0000\n",
      "('Pyrimidine metabolism PAO1', 'Pyrimidine deoxyribonuleotide biosynthesis, CDP/CTP => dCDP/dCTP,dTDP/dTTP'):\t0.0000\n",
      "('Bacterial secretion system PAO1', 'Sec (secretion) system'):\t0.0000\n",
      "('gamma-Hexachlorocyclohexane transport system', 'Phospholipid transport system'):\t0.0000\n",
      "('Limonene and pinene degradation PAO1', 'Lysine degradation PAO1'):\t0.0000\n",
      "('Tryptophan metabolism PAO1', 'Benzoate degradation PAO1'):\t0.0000\n",
      "('Xylene degradation PAO1', 'Benzoate degradation, benzoate => catechol / methylbenzoate => methylcatechol'):\t0.0000\n",
      "('Polyketide sugar unit biosynthesis PAO1', 'dTDP-L-rhamnose biosynthesis'):\t0.0000\n",
      "('Biosynthesis of unsaturated fatty acids PAO1', 'Pimeloyl-ACP biosynthesis, BioC-BioH pathway, malonyl-ACP => pimeloyl-ACP'):\t0.0000\n",
      "('Branched-chain amino acid transport system', 'Quorum sensing PAO1'):\t0.0000\n",
      "('Tryptophan metabolism PAO1', 'Valine, leucine and isoleucine degradation PAO1'):\t0.0000\n",
      "('Glycolysis / Gluconeogenesis PAO1', 'Glycolysis, core module involving three-carbon compounds'):\t0.0000\n",
      "('Biosynthesis of unsaturated fatty acids PAO1', 'Fatty acid biosynthesis, elongation'):\t0.0000\n",
      "('Oxidative phosphorylation PAO1', 'Cytochrome c oxidase, cbb3-type'):\t0.0000\n",
      "('Benzoate degradation PAO1', 'Valine, leucine and isoleucine degradation PAO1'):\t0.0000\n",
      "('2-Oxocarboxylic acid metabolism PAO1', 'Isoleucine biosynthesis, threonine => 2-oxobutanoate => isoleucine'):\t0.0000\n",
      "('Biosynthesis of unsaturated fatty acids PAO1', 'Fatty acid metabolism PAO1'):\t0.0000\n",
      "('Pantothenate and CoA biosynthesis PAO1', 'Pantothenate biosynthesis, valine/L-aspartate => pantothenate'):\t0.0000\n",
      "('Benzoate degradation PAO1', 'Fatty acid metabolism PAO1'):\t0.0000\n",
      "('Pyruvate metabolism PAO1', 'Pyruvate oxidation, pyruvate => acetyl-CoA'):\t0.0000\n",
      "('Synthesis and degradation of ketone bodies PAO1', 'Valine, leucine and isoleucine degradation PAO1'):\t0.0000\n",
      "('Geraniol degradation PAO1', 'beta-Oxidation'):\t0.0000\n",
      "('Gluconeogenesis, oxaloacetate => fructose-6P', 'Glycolysis, core module involving three-carbon compounds'):\t0.0000\n",
      "('C5-Branched dibasic acid metabolism PAO1', '2-Oxocarboxylic acid metabolism PAO1'):\t0.0000\n",
      "('Limonene and pinene degradation PAO1', 'beta-Alanine metabolism PAO1'):\t0.0000\n",
      "('Benzoate degradation PAO1', 'Butanoate metabolism PAO1'):\t0.0000\n",
      "('Limonene and pinene degradation PAO1', 'Tryptophan metabolism PAO1'):\t0.0000\n",
      "('Lysine degradation PAO1', 'Propanoate metabolism PAO1'):\t0.0000\n",
      "('Biotin metabolism PAO1', 'Biosynthesis of unsaturated fatty acids PAO1'):\t0.0000\n",
      "('Lysine degradation PAO1', 'Fatty acid metabolism PAO1'):\t0.0000\n",
      "('Pyrimidine metabolism PAO1', 'DNA polymerase III complex, bacteria'):\t0.0000\n",
      "('Arginine biosynthesis PAO1', 'Ornithine biosynthesis, glutamate => ornithine'):\t0.0000\n",
      "('Fatty acid metabolism PAO1', 'Fatty acid biosynthesis, initiation'):\t0.0000\n",
      "('2-Oxocarboxylic acid metabolism PAO1', 'Ornithine biosynthesis, glutamate => ornithine'):\t0.0000\n",
      "('Valine, leucine and isoleucine biosynthesis PAO1', 'Pantothenate and CoA biosynthesis PAO1'):\t0.0000\n",
      "('Fatty acid degradation PAO1', 'Limonene and pinene degradation PAO1'):\t0.0000\n",
      "('Fluorobenzoate degradation PAO1', 'Xylene degradation PAO1'):\t0.0000\n",
      "('Fluorobenzoate degradation PAO1', 'Benzoate degradation, benzoate => catechol / methylbenzoate => methylcatechol'):\t0.0000\n",
      "('Streptomycin biosynthesis PAO1', 'Polyketide sugar unit biosynthesis PAO1'):\t0.0000\n",
      "('Streptomycin biosynthesis PAO1', 'dTDP-L-rhamnose biosynthesis'):\t0.0000\n",
      "('Fluorobenzoate degradation PAO1', 'Degradation of aromatic compounds PAO1'):\t0.0000\n",
      "('Tyrosine metabolism PAO1', 'Degradation of aromatic compounds PAO1'):\t0.0000\n",
      "('Quorum sensing PAO1', 'Pyocyanine biosynthesis, chorismate => pyocyanine'):\t0.0000\n",
      "('Oxidative phosphorylation PAO1', 'F-type ATPase, prokaryotes and chloroplasts'):\t0.0000\n",
      "('Alanine, aspartate and glutamate metabolism PAO1', 'Nitrogen metabolism PAO1'):\t0.0000\n",
      "('beta-Lactam resistance PAO1', 'Multidrug resistance, efflux pump AcrAB-TolC/SmeDEF'):\t0.0000\n",
      "('Fatty acid degradation PAO1', 'Butanoate metabolism PAO1'):\t0.0000\n",
      "('Nitrogen metabolism PAO1', 'Dissimilatory nitrate reduction, nitrate => ammonia'):\t0.0000\n",
      "('Biosynthesis of unsaturated fatty acids PAO1', 'Fatty acid biosynthesis PAO1'):\t0.0000\n",
      "('Arginine biosynthesis PAO1', 'Nitrogen metabolism PAO1'):\t0.0000\n",
      "('Monobactam biosynthesis PAO1', 'Lysine biosynthesis, succinyl-DAP pathway, aspartate => lysine'):\t0.0000\n",
      "('Multidrug resistance, efflux pump MexAB-OprM', 'Multidrug resistance, efflux pump AcrAB-TolC/SmeDEF'):\t0.0000\n",
      "('Fatty acid degradation PAO1', 'Caprolactam degradation PAO1'):\t0.0000\n",
      "('Tryptophan metabolism PAO1', 'Butanoate metabolism PAO1'):\t0.0000\n",
      "('Histidine metabolism PAO1', 'Histidine degradation, histidine => N-formiminoglutamate => glutamate'):\t0.0000\n",
      "('Glycine, serine and threonine metabolism PAO1', 'Glyoxylate and dicarboxylate metabolism PAO1'):\t0.0000\n",
      "('Fluorobenzoate degradation PAO1', 'Benzoate degradation PAO1'):\t0.0000\n",
      "('Pentose phosphate pathway PAO1', 'Entner-Doudoroff pathway, glucose-6P => glyceraldehyde-3P + pyruvate'):\t0.0000\n",
      "('Degradation of aromatic compounds PAO1', 'Homoprotocatechuate degradation, homoprotocatechuate => 2-oxohept-3-enedioate'):\t0.0000\n",
      "('Sulfate-sulfur assimilation', 'Sulfate transport system'):\t0.0000\n",
      "('C5-Branched dibasic acid metabolism PAO1', 'Valine/isoleucine biosynthesis, pyruvate => valine / 2-oxobutanoate => isoleucine'):\t0.0000\n",
      "('Limonene and pinene degradation PAO1', 'Valine, leucine and isoleucine degradation PAO1'):\t0.0000\n",
      "('Tyrosine metabolism PAO1', 'Homoprotocatechuate degradation, homoprotocatechuate => 2-oxohept-3-enedioate'):\t0.0000\n",
      "('beta-Lactam resistance PAO1', 'Multidrug resistance, efflux pump MexAB-OprM'):\t0.0000\n",
      "('Limonene and pinene degradation PAO1', 'Caprolactam degradation PAO1'):\t0.0000\n"
     ]
    }
   ],
   "source": [
    "significant_edges = create_overlap_based_network(kegg_pathways_dict, alpha=0.05)\n",
    "print(\"Number of edges in the PAO1 KEGG overlap-based network: {0}\\n\".format(\n",
    "    len(significant_edges)))\n",
    "sorted_significant_edges = sorted(significant_edges, key=lambda tup: tup[1])\n",
    "\n",
    "show_N_edges = 150\n",
    "print(\"Pathway\\t\\tAdjusted p-value\")\n",
    "for edge, adjusted_pval in sorted_significant_edges[:show_N_edges]:\n",
    "    print(\"{0}:\\t{1:.4f}\".format(edge, adjusted_pval))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TCGA PID network summary statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pw0</th>\n",
       "      <th>pw1</th>\n",
       "      <th>weight</th>\n",
       "      <th>features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E2F</td>\n",
       "      <td>PLK1</td>\n",
       "      <td>6.543478</td>\n",
       "      <td>19.0 29.0 39.0 44.0 59.0 114.0 115.0 158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AURORA B</td>\n",
       "      <td>E2F</td>\n",
       "      <td>6.531936</td>\n",
       "      <td>19.0 29.0 39.0 44.0 59.0 114.0 115.0 158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AURORA B</td>\n",
       "      <td>PLK1</td>\n",
       "      <td>6.341347</td>\n",
       "      <td>19.0 29.0 39.0 44.0 59.0 114.0 115.0 158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E2F</td>\n",
       "      <td>FOXM1</td>\n",
       "      <td>5.974458</td>\n",
       "      <td>19.0 29.0 39.0 44.0 59.0 115.0 158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AURORA B</td>\n",
       "      <td>FOXM1</td>\n",
       "      <td>5.873026</td>\n",
       "      <td>19.0 29.0 39.0 44.0 59.0 115.0 158.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        pw0    pw1    weight                                    features\n",
       "0       E2F   PLK1  6.543478  19.0 29.0 39.0 44.0 59.0 114.0 115.0 158.0\n",
       "1  AURORA B    E2F  6.531936  19.0 29.0 39.0 44.0 59.0 114.0 115.0 158.0\n",
       "2  AURORA B   PLK1  6.341347  19.0 29.0 39.0 44.0 59.0 114.0 115.0 158.0\n",
       "3       E2F  FOXM1  5.974458        19.0 29.0 39.0 44.0 59.0 115.0 158.0\n",
       "4  AURORA B  FOXM1  5.873026        19.0 29.0 39.0 44.0 59.0 115.0 158.0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcga_pid_network = pd.read_table(tcga_pid_edges_file)\n",
    "tcga_pid_network.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# shorten the PID pathway names so that they match the ones\n",
    "# we use in the publication\n",
    "def shorten_tcga_pid(pathway_name):\n",
    "    REMOVE_SUFFIX = \"PATHWAY\"\n",
    "    pathway_short = None\n",
    "    split_on_underscores = pathway_name.split(\"_\")\n",
    "    if split_on_underscores[-1] == REMOVE_SUFFIX:\n",
    "        pathway_short = \" \".join(split_on_underscores[:-1])\n",
    "    else:\n",
    "        pathway_short = \" \".join(split_on_underscores)\n",
    "    return pathway_short\n",
    "\n",
    "pid_pathways_dict = load_pathway_definitions_file(\n",
    "    pid_pathways, shorten_tcga_pid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For the TCGA PID pathway-pathway co-occurrence network:\n",
      "Number of distinct pathways in the network: 89\n",
      "Number of edges in the network: 203\n",
      "Number of edges in the network where the 2 pathways share genes: 96 (average similarity by Jaccard Index: 0.05783)\n"
     ]
    }
   ],
   "source": [
    "pathways_in_pid_network = set(\n",
    "    tcga_pid_network[\"pw0\"].tolist() + tcga_pid_network[\"pw1\"].tolist())\n",
    "\n",
    "n_overlapping = 0\n",
    "edge_jaccards = []\n",
    "for index, row in tcga_pid_network.iterrows():\n",
    "    pw0_genes = pid_pathways_dict[row[\"pw0\"]] \n",
    "    pw1_genes = pid_pathways_dict[row[\"pw1\"]]\n",
    "    overlap = pw0_genes & pw1_genes\n",
    "    if overlap:\n",
    "        n_overlapping += 1\n",
    "    jaccard = float(len(overlap)) / (len(pw0_genes) + len(pw1_genes) - len(overlap))\n",
    "    edge_jaccards.append(jaccard)\n",
    "\n",
    "print(\"For the TCGA PID pathway-pathway co-occurrence network:\")\n",
    "print(\"Number of distinct pathways in the network: {0}\".format(\n",
    "    len(pathways_in_kegg_network)))\n",
    "print(\"Number of edges in the network: {0}\".format(\n",
    "    pao1_kegg_network.shape[0]))\n",
    "print(\"Number of edges in the network where the 2 pathways \"\n",
    "      \"share genes: {0} (average similarity by Jaccard Index: {1:.4})\".format(\n",
    "          n_overlapping, np.average(edge_jaccards)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison to a network based only on overlapping genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of edges in the PID overlap-based network: 3826\n",
      "\n",
      "Pathway\t\tAdjusted p-value\n",
      "('TCR', 'CD8 TCR'):\t0.0000\n",
      "('ERBB1 DOWNSTREAM', 'PDGFRB'):\t0.0000\n",
      "('IL8 CXCR2', 'IL8 CXCR1'):\t0.0000\n",
      "('AVB3 INTEGRIN', 'SYNDECAN 1'):\t0.0000\n",
      "('ENDOTHELIN', 'LYSOPHOSPHOLIPID'):\t0.0000\n",
      "('CDC42', 'RAC1'):\t0.0000\n",
      "('GMCSF', 'IL2 1PATHWAY'):\t0.0000\n",
      "('FCER1', 'BCR 5PATHWAY'):\t0.0000\n",
      "('GMCSF', 'IL3'):\t0.0000\n",
      "('CDC42', 'ERBB1 DOWNSTREAM'):\t0.0000\n",
      "('INTEGRIN1', 'INTEGRIN3'):\t0.0000\n",
      "('S1P S1P3', 'S1P S1P2'):\t0.0000\n",
      "('NFAT TFPATHWAY', 'TCR CALCIUM'):\t0.0000\n",
      "('ERBB1 DOWNSTREAM', 'ERBB2 ERBB3'):\t0.0000\n",
      "('ERBB1 RECEPTOR PROXIMAL', 'PDGFRB'):\t0.0000\n",
      "('SHP2', 'IL2 1PATHWAY'):\t0.0000\n",
      "('KIT', 'EPO'):\t0.0000\n",
      "('ER NONGENOMIC', 'THROMBIN PAR1'):\t0.0000\n",
      "('S1P S1P3', 'S1P S1P1'):\t0.0000\n",
      "('TXA2PATHWAY', 'IL8 CXCR1'):\t0.0000\n",
      "('IL2 PI3K', 'IL2 STAT5'):\t0.0000\n",
      "('TRAIL', 'FAS'):\t0.0000\n",
      "('LYSOPHOSPHOLIPID', 'ER NONGENOMIC'):\t0.0000\n",
      "('ENDOTHELIN', 'S1P S1P2'):\t0.0000\n",
      "('IL12 2PATHWAY', 'IL12 STAT4'):\t0.0000\n",
      "('MET', 'FAK'):\t0.0000\n",
      "('THROMBIN PAR4', 'THROMBIN PAR1'):\t0.0000\n",
      "('ERBB2 ERBB3', 'PDGFRB'):\t0.0000\n",
      "('FRA', 'AP1'):\t0.0000\n",
      "('HIV NEF', 'CASPASE'):\t0.0000\n",
      "('MET', 'PDGFRB'):\t0.0000\n",
      "('PDGFRB', 'FAK'):\t0.0000\n",
      "('FANCONI', 'ATR'):\t0.0000\n",
      "('MET', 'VEGFR1 2'):\t0.0000\n",
      "('CDC42', 'PDGFRB'):\t0.0000\n",
      "('MET', 'CDC42'):\t0.0000\n",
      "('S1P S1P3', 'S1P META'):\t0.0000\n",
      "('ENDOTHELIN', 'AR NONGENOMIC'):\t0.0000\n",
      "('IL2 1PATHWAY', 'IL2 STAT5'):\t0.0000\n",
      "('AR NONGENOMIC', 'CXCR3'):\t0.0000\n",
      "('GMCSF', 'ERBB2 ERBB3'):\t0.0000\n",
      "('VEGFR1', 'VEGFR1 2'):\t0.0000\n",
      "('TXA2PATHWAY', 'THROMBIN PAR1'):\t0.0000\n",
      "('NECTIN', 'ECADHERIN NASCENT AJ'):\t0.0000\n",
      "('IL2 1PATHWAY', 'PDGFRB'):\t0.0000\n",
      "('MET', 'RET'):\t0.0000\n",
      "('IL27', 'IL23'):\t0.0000\n",
      "('ERBB1 DOWNSTREAM', 'PI3K PLC TRK'):\t0.0000\n",
      "('INSULIN', 'MET'):\t0.0000\n",
      "('CONE', 'RHODOPSIN'):\t0.0000\n",
      "('MET', 'IGF1'):\t0.0000\n",
      "('ERBB1 RECEPTOR PROXIMAL', 'ERBB2 ERBB3'):\t0.0000\n",
      "('SHP2', 'TRKR'):\t0.0000\n",
      "('TXA2PATHWAY', 'IL8 CXCR2'):\t0.0000\n",
      "('ER NONGENOMIC', 'AR NONGENOMIC'):\t0.0000\n",
      "('IL12 2PATHWAY', 'IL23'):\t0.0000\n",
      "('INSULIN', 'IGF1'):\t0.0000\n",
      "('IL2 1PATHWAY', 'ERBB2 ERBB3'):\t0.0000\n",
      "('CERAMIDE', 'HIV NEF'):\t0.0000\n",
      "('FCER1', 'MET'):\t0.0000\n",
      "('ENDOTHELIN', 'S1P S1P3'):\t0.0000\n",
      "('ERBB1 RECEPTOR PROXIMAL', 'ERBB1 INTERNALIZATION'):\t0.0000\n",
      "('RET', 'PDGFRB'):\t0.0000\n",
      "('S1P META', 'S1P S1P2'):\t0.0000\n",
      "('NETRIN', 'FAK'):\t0.0000\n",
      "('TCPTP', 'ERBB1 RECEPTOR PROXIMAL'):\t0.0000\n",
      "('BETA CATENIN DEG', 'WNT CANONICAL'):\t0.0000\n",
      "('TCR RAS', 'CD8 TCR DOWNSTREAM'):\t0.0000\n",
      "('FCER1', 'GMCSF'):\t0.0000\n",
      "('ENDOTHELIN', 'THROMBIN PAR1'):\t0.0000\n",
      "('PI3KCI', 'PDGFRB'):\t0.0000\n",
      "('PDGFRB', 'RAC1'):\t0.0000\n",
      "('RET', 'FAK'):\t0.0000\n",
      "('ERBB2 ERBB3', 'CXCR3'):\t0.0000\n",
      "('EPHB FWD', 'RET'):\t0.0000\n",
      "('EPHB FWD', 'FAK'):\t0.0000\n",
      "('INTEGRIN4', 'A6B1 A6B4 INTEGRIN'):\t0.0000\n",
      "('MET', 'ERBB1 DOWNSTREAM'):\t0.0000\n",
      "('MET', 'TRKR'):\t0.0000\n",
      "('GMCSF', 'PDGFRB'):\t0.0000\n",
      "('TXA2PATHWAY', 'CXCR4'):\t0.0000\n",
      "('FCER1', 'IL2 1PATHWAY'):\t0.0000\n",
      "('PTP1B', 'PDGFRB'):\t0.0000\n",
      "('ERBB1 RECEPTOR PROXIMAL', 'ERBB1 DOWNSTREAM'):\t0.0000\n",
      "('MTOR 4PATHWAY', 'ERBB1 DOWNSTREAM'):\t0.0000\n",
      "('MET', 'ERBB2 ERBB3'):\t0.0000\n",
      "('CXCR4', 'PDGFRB'):\t0.0000\n",
      "('ENDOTHELIN', 'ER NONGENOMIC'):\t0.0000\n",
      "('EPHB FWD', 'ERBB1 RECEPTOR PROXIMAL'):\t0.0000\n",
      "('IL12 2PATHWAY', 'CD8 TCR DOWNSTREAM'):\t0.0000\n",
      "('HDAC CLASSII', 'HDAC CLASSI'):\t0.0000\n",
      "('MET', 'EPHB FWD'):\t0.0000\n",
      "('CDC42 REG', 'RAC1 REG'):\t0.0000\n",
      "('ERBB2 ERBB3', 'ERBB1 INTERNALIZATION'):\t0.0000\n",
      "('TNF', 'CERAMIDE'):\t0.0000\n",
      "('CXCR4', 'CXCR3'):\t0.0000\n",
      "('PTP1B', 'TCPTP'):\t0.0000\n",
      "('LYSOPHOSPHOLIPID', 'S1P S1P2'):\t0.0000\n",
      "('RET', 'IGF1'):\t0.0000\n",
      "('IL2 1PATHWAY', 'IL2 PI3K'):\t0.0000\n",
      "('MET', 'ERBB1 RECEPTOR PROXIMAL'):\t0.0000\n",
      "('CXCR4', 'PI3KCI'):\t0.0000\n",
      "('GMCSF', 'IL5'):\t0.0000\n",
      "('ERBB1 DOWNSTREAM', 'CXCR3'):\t0.0000\n",
      "('INTEGRIN1', 'AVB3 INTEGRIN'):\t0.0000\n",
      "('S1P S1P4', 'S1P S1P2'):\t0.0000\n",
      "('GMCSF', 'KIT'):\t0.0000\n",
      "('FCER1', 'PDGFRB'):\t0.0000\n",
      "('IL27', 'IL12 2PATHWAY'):\t0.0000\n",
      "('LYSOPHOSPHOLIPID', 'THROMBIN PAR1'):\t0.0000\n",
      "('MET', 'NEPHRIN NEPH1'):\t0.0000\n",
      "('ERBB1 DOWNSTREAM', 'MAPK TRK'):\t0.0000\n",
      "('PDGFRB', 'VEGFR1 2'):\t0.0000\n",
      "('LYSOPHOSPHOLIPID', 'AR NONGENOMIC'):\t0.0000\n",
      "('TNF', 'HIV NEF'):\t0.0000\n",
      "('EPHB FWD', 'TRKR'):\t0.0000\n",
      "('ANGIOPOIETIN RECEPTOR', 'FAK'):\t0.0000\n",
      "('IL5', 'IL3'):\t0.0000\n",
      "('INTEGRIN1', 'SYNDECAN 1'):\t0.0000\n",
      "('CD8 TCR DOWNSTREAM', 'IL12 STAT4'):\t0.0000\n",
      "('FANCONI', 'BARD1'):\t0.0000\n",
      "('IL2 1PATHWAY', 'CD8 TCR DOWNSTREAM'):\t0.0000\n",
      "('ERBB2 ERBB3', 'NEPHRIN NEPH1'):\t0.0000\n",
      "('ENDOTHELIN', 'TXA2PATHWAY'):\t0.0000\n",
      "('GMCSF', 'SHP2'):\t0.0000\n",
      "('RET', 'ERBB1 RECEPTOR PROXIMAL'):\t0.0000\n",
      "('S1P S1P3', 'S1P S1P4'):\t0.0000\n",
      "('ER NONGENOMIC', 'CXCR3'):\t0.0000\n",
      "('ERBB1 DOWNSTREAM', 'RAC1'):\t0.0000\n",
      "('BCR 5PATHWAY', 'TCR'):\t0.0000\n",
      "('PDGFRB', 'PI3K PLC TRK'):\t0.0000\n",
      "('INTEGRIN3', 'AVB3 INTEGRIN'):\t0.0000\n",
      "('FCER1', 'KIT'):\t0.0000\n",
      "('IL8 CXCR2', 'THROMBIN PAR1'):\t0.0000\n",
      "('LYSOPHOSPHOLIPID', 'S1P S1P3'):\t0.0000\n",
      "('ERBB2 ERBB3', 'KIT'):\t0.0000\n",
      "('ERBB1 RECEPTOR PROXIMAL', 'CXCR3'):\t0.0000\n",
      "('ER NONGENOMIC', 'S1P S1P3'):\t0.0000\n",
      "('IGF1', 'KIT'):\t0.0000\n",
      "('THROMBIN PAR1', 'IL8 CXCR1'):\t0.0000\n",
      "('TCR RAS', 'MAPK TRK'):\t0.0000\n",
      "('IL2 1PATHWAY', 'KIT'):\t0.0000\n",
      "('RET', 'ANGIOPOIETIN RECEPTOR'):\t0.0000\n",
      "('S1P S1P4', 'S1P S1P1'):\t0.0000\n",
      "('S1P S1P4', 'S1P META'):\t0.0000\n",
      "('INSULIN GLUCOSE', 'PI3KCI AKT'):\t0.0000\n",
      "('RET', 'TRKR'):\t0.0000\n",
      "('TCPTP', 'PDGFRB'):\t0.0000\n",
      "('CXCR4', 'AR NONGENOMIC'):\t0.0000\n",
      "('MET', 'NETRIN'):\t0.0000\n"
     ]
    }
   ],
   "source": [
    "significant_edges = create_overlap_based_network(pid_pathways_dict, alpha=0.05)\n",
    "print(\"Number of edges in the PID overlap-based network: {0}\\n\".format(\n",
    "    len(significant_edges)))\n",
    "sorted_significant_edges = sorted(significant_edges, key=lambda tup: tup[1])\n",
    "\n",
    "show_N_edges = 150\n",
    "print(\"Pathway\\t\\tAdjusted p-value\")\n",
    "for edge, adjusted_pval in sorted_significant_edges[:show_N_edges]:\n",
    "    print(\"{0}:\\t{1:.4f}\".format(edge, adjusted_pval))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
